File: src/main/java/core/procedures/vectors/Vec.java
package core.procedures.vectors;

import core.exceptions.WrongTypeException;
import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.MutableVector;

import java.util.Collection;

public final class Vec extends AFn {

  public Vec() {
    super(new FnArgsBuilder().min(1).max(1).build());
  }

  @Override
  public String getName() {
    return "vec";
  }

  @Override
  public MutableVector apply1(Object arg) {
    if (arg instanceof Collection) {
      return new MutableVector(((Collection)arg).toArray());
    }
    if (arg instanceof CharSequence) {
      int size = ((CharSequence) arg).length();
      MutableVector vector = new MutableVector(size, null);
      for (int i = 0; i < size; i++) {
        vector.set(i, ((CharSequence)arg).charAt(i));
      }
      return vector;
    }
    throw new WrongTypeException(getName(), "List or Vector or String", arg);
  }
}

File: src/main/java/core/procedures/vectors/Vec.kt
package core.procedures.vectors

import core.exceptions.WrongTypeException
import core.procedures.AFn
import core.scm.MutableVector

class Vec : AFn<Any?, MutableVector>(name = "vec", isPure = true, minArgs = 1, maxArgs = 1) {

    override operator fun invoke(arg: Any?) = when (arg) {
        is Collection<*> -> MutableVector(arg.toTypedArray())
        is CharSequence -> MutableVector(arg.length, null).apply {
            for (i in 0 until arg.length) {
                this[i] = arg[i]
            }
        }
        else -> throw WrongTypeException(name, "List or Vector or String", arg)
    }
}
